Add typing indicators for streams.

chdinesh1089 opened the pull request:
Fixes #11152

This is almost done. Things left to be done are

- [x] Fix and add few frontend tests.
- [x] Document the new arguments `stream_id` and `topic` for `api/v1/typing` in `zerver/openapi/zulip.yaml`.
- [x] Send typing notification only if the topic already exists as we want to show typing notification only in a topic narrow.

Sending a request to streams with 0 or 1 subscriber(only sender) won't throw any error. Do we want to return any error for that?

chdinesh1089 commented:
@showell I see this in the issue page.
>  This is a non-trivial feature, and you'll want to discuss performance on #backend.

Could you do a review and tell if I should open a thread for this in #backend

showell commented on zerver/lib/actions.py:
```diff
@@ -1888,20 +1888,27 @@ def do_send_typing_notification(
         realm: Realm,
```
I would prefer to just have a separate action function for stream notifications, so that you don't need the optional parameters and the conditional logic.  Two simple functions are better than one complex one, and we're not really sharing any useful logic here.

I would also consider just having a new `type` (no pun intended) for the event with the value `stream_typing`, so that the documentation piece is easier here (again, no Union types would be needed).

chdinesh1089 replied:
Yes, having a different type would make it simpler and separating the function will make it seem less complex. Will make the required changes.

(Thought I'd make the changes and ask for another review soon but not I'm getting a chance to work on this as most of my time is going into academics. Leaving an update since it's been 2weeks..:)

timabbott replied:
Yeah, I think the separate function makes sense, with a bit of preparatory refactoring to avoid duplicating code.

chdinesh1089 replied:
Separated the function. There wasn't much duplicate code. 

Also added `type` `stream_typing`.

timabbott commented on zerver/views/typing.py:
```diff
@@ -14,6 +15,21 @@ def send_notification_backend(
+        return json_error(_("Insufficient arguments. Should have 'to' or both 'stream_id' and 'topic'."))
+
+    if all([notification_to, to_stream, to_topic]):
+        return json_error(_("All 'to', 'stream_id', and  'topic' at once are not accepted"))
```
I think this should probably be the same error message as the previous one, to avoid making our translators do extra strings for no reason.

chdinesh1089 replied:
Changed all errors regarding this to "Bad arguments. Should have 'to' or both 'stream_id' and 'topic'."

timabbott commented on zerver/lib/actions.py:
```diff
@@ -1911,16 +1918,31 @@ def do_send_typing_notification(
+        raise JsonableError(_('Invalid \'op\' value (should be start or stop)'))
+
+    try:
+        get_stream_by_id_in_realm(stream_id, realm)
```
Let's use the `access_stream_...` methods, called from the view function -- that's our standard approach to security.

I think it's also wise to do a preparatory refactoring commit to move the `start|stop` validation into the view function -- I think you can do it in `REQ` with `check_string_in`.

chdinesh1089 replied:
Changed to do validation using `access_stream_by_id`.

Yes, It can be done using `REQ` with `check_string_in`. Added a preparatory commit. One thing to notice here is doing it in `REQ` would raise error just saying `Invalid op`but we'll miss the verbose error `Invalid 'op' value (should be start or stop` we were returning.

chdinesh1089 commented:
Sorry for the delay. Haven't been getting time due to academics. Our teachers are tending to give more assignments as everything is still online :grimacing: 

 Made the changes based on reviews. I'll try to do the other work i.e adding frontend tests and documenting in the weekend.

mateuszmandera commented on zerver/views/typing.py:
```diff
@@ -15,6 +16,23 @@ def send_notification_backend(
+
+    bad_arguments_error = _("Bad arguments. Should have 'to' or both 'stream_id' and 'topic'.")
+
+    if not any([notification_to, to_stream, to_topic]):
+        return json_error(bad_arguments_error)
+
+    if all([notification_to, to_stream, to_topic]):
+        return json_error(bad_arguments_error)
+
+    if notification_to and not (to_stream or to_topic):
+        check_send_typing_notification(user_profile, notification_to, operator)
+    elif to_stream and to_topic:
+        check_send_stream_typing_notification(user_profile, operator, to_stream, to_topic)
+    else:
+        return json_error(bad_arguments_error)
```
Perhaps it would  be cleaner to write this somewhat like this?
```
    if any(
        notification_to and (to_stream or to_topic),
        not notification_to and not(to_stream and to_topic)
    ):
        return json_error(bad_arguments_error)

    if notification_to:
        check_send_typing_notification(user_profile, notification_to, operator)
    else:
        check_send_stream_typing_notification(user_profile, operator, to_stream, to_topic)
```

chdinesh1089 commented on frontend_tests/node_tests/typing_status.js:
```diff
@@ -263,6 +264,7 @@ run_test("basics", () => {
+    compose_state.get_message_type = () => "private";
     compose_pm_pill.get_user_ids_string = () => "1,2,3";
     typing_status.state.current_recipient = typing.get_recipient();
 
```
I haven't added/modified tests here as it's intended to test `typing_status.js` to which this PR makes no changes. But we make use of this module to send start/stop requests from client which only extends the possible values for `status.current_recipient` to also be an object consisting `stream_id` and `topic`. since we only check if current recipient is equal to previous recipient, adding tests to include recipient format of object:`stream_id, topic` will be redundant and won't add much value I think.

Will add few tests to also include that format for `current_recipient` if you think it's necessary.

chdinesh1089 commented:
Do we want to show "several people are typing..." only for streams or for PMs too if more than 3 users are typing?

chdinesh1089 commented:
How do I fix the API test failure? I'm unable to figure out a solution. Here's the issue:
We now added to extra params(`stream_id`, `topic` that could be sent to `json/typing`. The `test_curl_examples` is failing as it's generating an example consisting of all three parameters which gives an error according to the actual code but the test is asserting it with a success message.

Tried adding this in zulip.yaml, but it still didn't work.
```
        "400":
          description: Bad request.
          content:
            application/json:
              schema:
                allOf:
                  - $ref: "#/components/schemas/JsonError"
                  - example:
                      {
                        "code": "BAD_REQUEST",
                        "msg": "Bad arguments. Should have 'to' or both 'stream_id' and 'topic'.",
                        "result": "error",
                      }
```
I think we should still have this. will push it once we figure out a fix for the failure.

chdinesh1089 commented on zerver/openapi/python_examples.py:
```diff
@@ -1081,6 +1081,7 @@ def set_typing_status(client: Client) -> None:
         "to": [user_id1, user_id2],
     }
     result = client.set_typing_status(request)
+
```
Added this space(and below) intentionally to make it look good in API docs with each of start, stop separated as a different chunk.

timabbott commented on static/js/typing.js:
```diff
@@ -92,6 +93,9 @@ export function get_recipient() {
         const topic = compose_state.topic();
+        if (!stream_topic_history.get_recent_topic_names(stream_id).includes(topic)) {
+            return null; // Send notification only if topic already exists.
+        }
```
I don't think this last change is correct or necessary -- one can narrow to a topic that doesn't exist, and if you do so at the same time as another user who is typing a message to that proposed new topic (E.g. someone said "Let's continue here" with a link), not adding this check will result in typing notifications working for those users.

(Well, we'd need to test the display UI in the empty narrow view for the topic, but that's it).

chdinesh1089 replied:
Look like this in the empty narrow view: ![image](https://user-images.githubusercontent.com/44665669/113619636-db086580-9676-11eb-805c-4a7687b540a7.png)

timabbott commented on static/js/typing_events.js:
```diff
@@ -54,10 +58,18 @@ function get_users_typing_for_narrow() {
+            $("#typing_notifications").html(render_typing_notifications({users: users_typing}));
+        } else {
+            $("#typing_notifications").html(
+                "<li class='typing_notification'>Several people are typing...</li>",
```
This string should be tagged for translation.  

Though I think cleaner would be to just move this conditional into the HTML template rendered by `render_typing_notifications`.

timabbott replied:
I think if you clean up this commit, we can probably just merge this at the beginning of the branch as a small feature / refactoring.

Oh, and use the unicode `...` character; it looks nicer than 3 dots.

chdinesh1089 replied:
Do you mean this: https://en.wikipedia.org/wiki/Ellipsis ? It seems to have the dots closer. `__user__ is typing...` too had 3 dots. That too should be changed I think?(perhaps in a different commit)

Since I wasn't sure if you meant the ellipsis character and it looked odd to me, I left the that as three dots.

I cleaned up the commit, moved the conditional to the `.hbs` template, and added it for translation.

chdinesh1089 replied:
Replaced the 3-dots with the Unicode … character. It looked odd in the code editor, so I was confused. 

Yes, the … character looks nicer :)
Comparison: (1st is unicode … char, and the second is our previous 3 dots.)
![image](https://user-images.githubusercontent.com/44665669/113758314-529acb00-9731-11eb-95bc-3638f4324f44.png)

timabbott commented on static/js/typing_data.js:
```diff
@@ -21,8 +21,7 @@ function get_key(group) {
 
-export function add_typist(group, typist) {
-    const key = get_key(group);
+function update_typist_dct(key, typist) {
```
This refactor to the `update_typist_dct` interface would be nice if extracted as a prep commit.

I'd also call this `set_typing_state(key, typist)`.

chdinesh1089 replied:
`add_typist` works for both if `key` is extracted out. So we don't need this refactor now!

timabbott commented on static/js/typing_data.js:
```diff
@@ -64,20 +82,40 @@ export function get_all_pms_typists() {
+}
+
+export function clear_streams_inbound_timer(stream_id, topic) {
+    const key = JSON.stringify({stream_id, topic});
```
We're going to want this to be a `get_topic_key(stream_id, topic)` -- topics are case-insensitive, and we'll want to implement that in only one place.

chdinesh1089 replied:
Added a `get_topic_key`.

timabbott commented on static/js/typing_data.js:
```diff
@@ -64,20 +82,40 @@ export function get_all_pms_typists() {
+    clear_streams_inbound_timer(stream_id, topic);
     const timer = setTimeout(callback, delay);
     inbound_timer_dict.set(key, timer);
 }
```
I'm not sure you actually need to duplicate these functions -- can't we just have whatever is calling them fetch the key and call `kickstart_inbound_timer(key, delay, callback)`?

chdinesh1089 replied:
Yes, agreed, that's a lot of duplication. Changed to fetch the key and call `kickstart_inbound_timer(key, delay, callback)` as suggested.

timabbott commented:
Having read the frontend somewhat now, I think we're taking the wrong strategy with duplicating these functions; it looks like fundamentally, the only thing that's different between the PMs and stream/topic implementations is how we construct the keys for our dictionary, and we should be able to do that more cleanly with just a couple conditionals rather than duplicating 80% of some of these modules.

timabbott commented:
Thanks for working on this @chdinesh1089!  I merged a handful of prep commits as the series ending with 08116a17b0dfec59b5eba7e97facb13fcad56309; note the way I split the refactoring up a bit more than you had.  

I started a conversation in https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/stream.20typing.20notifications/near/1149619 on the API design for this, but my belief is that we should plan to do the `message_type` variable strategy we use for sending messages, since this is intended to be a parallel API.  I think it's reasonable to just have `to` be a `List[int]` unconditionally -- where for type=stream messages it needs to have a single element in it.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2263,6 +2265,38 @@ def check_send_typing_notification(sender: UserProfile, user_ids: List[int], ope
+
+    access_stream_by_id(sender, stream_id)  # Does the checking
+
+    do_send_stream_typing_notification(realm, sender, operator, stream_id, topic)
```
I don't think we need this function -- we should just do `access_stream_by_id` in the views code (as is our usual policy) and then pass the `Stream` object to `do_send_stream_typing_notification`. (which can then just access `stream.id`).

chdinesh1089 replied:
We don't need a `Stream` object anywhere. Moving that `access_stream_by_id` check to the view function and removing this function.

chdinesh1089 commented:
Reworked the backend changes as suggested. Wasn't able to do the frontend changes. (was taking more time than I anticipated) I'll do those after exams or sometime in the middle if I get time.

chdinesh1089 commented:
Done with the frontend changes too! This is ready for another review.

timabbott commented:
I merged 3 commits, and pushed the rest back here; will post comments on the next few.

timabbott commented on templates/zerver/api/set-typing-status.md:
```diff
@@ -17,7 +17,9 @@ More examples and documentation can be found [here](https://github.com/zulip/zul
 {tab|curl}
 
-{generate_code_example(curl)|/typing:post|example}
+You may set typing status for private message or group private messages as follows:
```
I think we can just have the example without text.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -9258,14 +9365,34 @@ paths:
+        - name: topic
+          in: query
+          description: |
+            Topic to which message is being typed. Valid only for type 'stream'.
```
Should say "Required for the "stream" type" or something to that effect.

chdinesh1089 replied:
The description now is:
>Topic to which message is being typed. Required for the 'stream' type.
            Ignored in case of 'private' type.

timabbott commented on zerver/views/typing.py:
```diff
@@ -21,10 +22,27 @@ def send_notification_backend(
+            return json_error(_("Multiple stream IDs not supported"))
+
+        if topic is None:
+            return json_error(_("Missing parameter 'topic'"))
```
To what extent do these error strings match the errors the `send_message` API gives for the same parameter error?  Ideally, we'd try to match that to avoid making translators do extra work for this endpoint.

chdinesh1089 replied:
For the empty `to` list I think it'd be better to go with `Empty 'to' list` since the `send_message` says `Message must have recipients` for private and `Missing stream` for stream types.

Changed the other two to `Cannot send to multiple streams` and `Missing topic` which match with those of `send_message`.

timabbott commented on zerver/tests/test_typing.py:
```diff
@@ -290,3 +344,75 @@ def test_stop_to_another_user(self) -> None:
+            result = self.api_post(sender, "/api/v1/typing", params)
+
+        self.assert_json_success(result)
+        self.assertEqual(len(events), 1)
```
Can you do query count assertions in these new tests as well?  They are somewhat important given that this is a high-volume endpoint.

chdinesh1089 replied:
Sure, added those. It isn't clear to me how those help so I skipped them at the time of writing those thinking I'd ask about them later but forgot. Could you provide more data/ pointers to understand how query count assertions in tests help? 

To make sure what I had done is right: I just checked the length of `queries` with a `print` and added an assert to that value. Is it what is to be done, or there's something else for me to do to make sure 12 queries are run?

timabbott replied:
Yeah, that's the right way to do it.  It's worth printing the queries once nicely one-per-line and pasting the result here, so we can confirm it's not broken.

chdinesh1089 replied:
My previous code was inefficient in fetching users IDs. It was running a query of each user, just to get user ID. Changed it now to use an existing function `get_user_ids_for_streams`. 

These are the queries: (Not sure why there's a client = Mobile query, I presume it's somehow related to `self.api_post` of tests)

0 SELECT "zerver_userprofile"."id", "zerver_userprofile"."password", "zerver_userprofile"."last_login", "zerver_userprofile"."is_superuser", "zerver_userprofile"."delivery_email", "zerver_userprofile"."email", "zerver_userprofile"."realm_id", "zerver_userprofile"."recipient_id", "zerver_userprofile"."full_name", "zerver_userprofile"."date_joined", "zerver_userprofile"."tos_version", "zerver_userprofile"."api_key", "zerver_userprofile"."is_staff", "zerver_userprofile"."is_active", "zerver_userprofile"."is_billing_admin", "zerver_userprofile"."is_bot", "zerver_userprofile"."bot_type", "zerver_userprofile"."bot_owner_id", "zerver_userprofile"."role", "zerver_userprofile"."long_term_idle", "zerver_userprofile"."last_active_message_id", "zerver_userprofile"."is_mirror_dummy", "zerver_userprofile"."can_forge_sender", "zerver_userprofile"."can_create_users", "zerver_userprofile"."enable_stream_desktop_notifications", "zerver_userprofile"."enable_stream_email_notifications", "zerver_userprofile"."enable_stream_push_notifications", "zerver_userprofile"."enable_stream_audible_notifications", "zerver_userprofile"."notification_sound", "zerver_userprofile"."wildcard_mentions_notify", "zerver_userprofile"."enable_desktop_notifications", "zerver_userprofile"."pm_content_in_desktop_notifications", "zerver_userprofile"."enable_sounds", "zerver_userprofile"."enable_offline_email_notifications", "zerver_userprofile"."message_content_in_email_notifications", "zerver_userprofile"."enable_offline_push_notifications", "zerver_userprofile"."enable_online_push_notifications", "zerver_userprofile"."desktop_icon_count_display", "zerver_userprofile"."enable_digest_emails", "zerver_userprofile"."enable_login_emails", "zerver_userprofile"."realm_name_in_notifications", "zerver_userprofile"."presence_enabled", "zerver_userprofile"."last_reminder", "zerver_userprofile"."rate_limits", "zerver_userprofile"."default_sending_stream_id", "zerver_userprofile"."default_events_register_stream_id", "zerver_userprofile"."default_all_public_streams", "zerver_userprofile"."enter_sends", "zerver_userprofile"."left_side_userlist", "zerver_userprofile"."default_language", "zerver_userprofile"."default_view", "zerver_userprofile"."dense_mode", "zerver_userprofile"."fluid_layout_width", "zerver_userprofile"."high_contrast_mode", "zerver_userprofile"."translate_emoticons", "zerver_userprofile"."twenty_four_hour_time", "zerver_userprofile"."starred_message_counts", "zerver_userprofile"."color_scheme", "zerver_userprofile"."demote_inactive_streams", "zerver_userprofile"."timezone", "zerver_userprofile"."emojiset", "zerver_userprofile"."avatar_source", "zerver_userprofile"."avatar_version", "zerver_userprofile"."avatar_hash", "zerver_userprofile"."tutorial_status", "zerver_userprofile"."onboarding_steps", "zerver_userprofile"."zoom_token", "zerver_realm"."id", "zerver_realm"."name", "zerver_realm"."description", "zerver_realm"."string_id", "zerver_realm"."date_created", "zerver_realm"."deactivated", "zerver_realm"."deactivated_redirect", "zerver_realm"."emails_restricted_to_domains", "zerver_realm"."invite_required", "zerver_realm"."max_invites", "zerver_realm"."disallow_disposable_email_addresses", "zerver_realm"."authentication_methods", "zerver_realm"."inline_image_preview", "zerver_realm"."inline_url_embed_preview", "zerver_realm"."digest_emails_enabled", "zerver_realm"."digest_weekday", "zerver_realm"."send_welcome_emails", "zerver_realm"."message_content_allowed_in_email_notifications", "zerver_realm"."mandatory_topics", "zerver_realm"."add_emoji_by_admins_only", "zerver_realm"."name_changes_disabled", "zerver_realm"."email_changes_disabled", "zerver_realm"."avatar_changes_disabled", "zerver_realm"."create_stream_policy", "zerver_realm"."invite_to_realm_policy", "zerver_realm"."invite_to_stream_policy", "zerver_realm"."move_messages_between_streams_policy", "zerver_realm"."user_group_edit_policy", "zerver_realm"."private_message_policy", "zerver_realm"."wildcard_mention_policy", "zerver_realm"."email_address_visibility", "zerver_realm"."waiting_period_threshold", "zerver_realm"."allow_message_deleting", "zerver_realm"."message_content_delete_limit_seconds", "zerver_realm"."allow_message_editing", "zerver_realm"."message_content_edit_limit_seconds", "zerver_realm"."allow_edit_history", "zerver_realm"."allow_community_topic_editing", "zerver_realm"."default_twenty_four_hour_time", "zerver_realm"."default_language", "zerver_realm"."notifications_stream_id", "zerver_realm"."signup_notifications_stream_id", "zerver_realm"."message_retention_days", "zerver_realm"."message_visibility_limit", "zerver_realm"."first_visible_message_id", "zerver_realm"."org_type", "zerver_realm"."plan_type", "zerver_realm"."bot_creation_policy", "zerver_realm"."upload_quota_gb", "zerver_realm"."video_chat_provider", "zerver_realm"."giphy_rating", "zerver_realm"."default_code_block_language", "zerver_realm"."icon_source", "zerver_realm"."icon_version", "zerver_realm"."logo_source", "zerver_realm"."logo_version", "zerver_realm"."night_logo_source", "zerver_realm"."night_logo_version" FROM "zerver_userprofile" INNER JOIN "zerver_realm" ON ("zerver_userprofile"."realm_id" = "zerver_realm"."id") WHERE "zerver_userprofile"."api_key" = 'B7mRHbe3e8yHJOOuBKk6cf5fWnO2Bsj6' LIMIT 21

1 SELECT "zerver_client"."id", "zerver_client"."name" FROM "zerver_client" WHERE "zerver_client"."name" = 'ZulipMobile' LIMIT 21

2 SELECT "zerver_stream"."id", "zerver_stream"."name", "zerver_stream"."realm_id", "zerver_stream"."date_created", "zerver_stream"."deactivated", "zerver_stream"."description", "zerver_stream"."rendered_description", "zerver_stream"."recipient_id", "zerver_stream"."invite_only", "zerver_stream"."history_public_to_subscribers", "zerver_stream"."is_web_public", "zerver_stream"."stream_post_policy", "zerver_stream"."is_in_zephyr_realm", "zerver_stream"."email_token", "zerver_stream"."message_retention_days", "zerver_stream"."first_message_id", "zerver_realm"."id", "zerver_realm"."name", "zerver_realm"."description", "zerver_realm"."string_id", "zerver_realm"."date_created", "zerver_realm"."deactivated", "zerver_realm"."deactivated_redirect", "zerver_realm"."emails_restricted_to_domains", "zerver_realm"."invite_required", "zerver_realm"."max_invites", "zerver_realm"."disallow_disposable_email_addresses", "zerver_realm"."authentication_methods", "zerver_realm"."inline_image_preview", "zerver_realm"."inline_url_embed_preview", "zerver_realm"."digest_emails_enabled", "zerver_realm"."digest_weekday", "zerver_realm"."send_welcome_emails", "zerver_realm"."message_content_allowed_in_email_notifications", "zerver_realm"."mandatory_topics", "zerver_realm"."add_emoji_by_admins_only", "zerver_realm"."name_changes_disabled", "zerver_realm"."email_changes_disabled", "zerver_realm"."avatar_changes_disabled", "zerver_realm"."create_stream_policy", "zerver_realm"."invite_to_realm_policy", "zerver_realm"."invite_to_stream_policy", "zerver_realm"."move_messages_between_streams_policy", "zerver_realm"."user_group_edit_policy", "zerver_realm"."private_message_policy", "zerver_realm"."wildcard_mention_policy", "zerver_realm"."email_address_visibility", "zerver_realm"."waiting_period_threshold", "zerver_realm"."allow_message_deleting", "zerver_realm"."message_content_delete_limit_seconds", "zerver_realm"."allow_message_editing", "zerver_realm"."message_content_edit_limit_seconds", "zerver_realm"."allow_edit_history", "zerver_realm"."allow_community_topic_editing", "zerver_realm"."default_twenty_four_hour_time", "zerver_realm"."default_language", "zerver_realm"."notifications_stream_id", "zerver_realm"."signup_notifications_stream_id", "zerver_realm"."message_retention_days", "zerver_realm"."message_visibility_limit", "zerver_realm"."first_visible_message_id", "zerver_realm"."org_type", "zerver_realm"."plan_type", "zerver_realm"."bot_creation_policy", "zerver_realm"."upload_quota_gb", "zerver_realm"."video_chat_provider", "zerver_realm"."giphy_rating", "zerver_realm"."default_code_block_language", "zerver_realm"."icon_source", "zerver_realm"."icon_version", "zerver_realm"."logo_source", "zerver_realm"."logo_version", "zerver_realm"."night_logo_source", "zerver_realm"."night_logo_version" FROM "zerver_stream" INNER JOIN "zerver_realm" ON ("zerver_stream"."realm_id" = "zerver_realm"."id") WHERE ("zerver_stream"."id" = 1 AND "zerver_stream"."realm_id" = 2) LIMIT 21

3 SELECT "zerver_subscription"."id", "zerver_subscription"."user_profile_id", "zerver_subscription"."recipient_id", "zerver_subscription"."active", "zerver_subscription"."is_user_active", "zerver_subscription"."role", "zerver_subscription"."is_muted", "zerver_subscription"."color", "zerver_subscription"."pin_to_top", "zerver_subscription"."desktop_notifications", "zerver_subscription"."audible_notifications", "zerver_subscription"."push_notifications", "zerver_subscription"."email_notifications", "zerver_subscription"."wildcard_mentions_notify" FROM "zerver_subscription" WHERE ("zerver_subscription"."active" AND "zerver_subscription"."recipient_id" = 120 AND "zerver_subscription"."user_profile_id" = 10) LIMIT 21

4 SELECT "zerver_recipient"."type_id", "zerver_subscription"."user_profile_id" FROM "zerver_subscription" INNER JOIN "zerver_recipient" ON ("zerver_subscription"."recipient_id" = "zerver_recipient"."id") INNER JOIN "zerver_userprofile" ON ("zerver_subscription"."user_profile_id" = "zerver_userprofile"."id") WHERE ("zerver_subscription"."active" AND "zerver_recipient"."type" = 2 AND "zerver_recipient"."type_id" IN (1) AND "zerver_userprofile"."is_active") ORDER BY "zerver_recipient"."type_id" ASC

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -9258,14 +9365,34 @@ paths:
+                  - example:
+                      {
+                        "code": "BAD_REQUEST",
+                        "msg": "Bad arguments. Should have 'to' or both 'stream_id' and 'topic'.",
```
This is out of date, I think.

chdinesh1089 replied:
yes, changed.

timabbott commented on zerver/openapi/python_examples.py:
```diff
@@ -1162,10 +1163,47 @@ def set_typing_status(client: Client) -> None:
+
+    validate_against_openapi_schema(result, "/typing", "post", "200")
+
+    # client.delete_stream(stream_id)
```
Is this comment intended to be here?

chdinesh1089 replied:
Removed, was experimenting in the initial versions and forgot to remove.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2289,6 +2290,28 @@ def check_send_typing_notification(sender: UserProfile, user_ids: List[int], ope
+    sender_dict = {"user_id": sender.id, "email": sender.email}
+
+    event = dict(
+        type="stream_typing",
```
I'd like to have this be a `typing` event, not a `stream_typing` event.  To do that, we need to do the following:
* Reorder this branch so that the frontend work to at least make `server_events.js` not crash when seeing a `typing` event with a `stream_id` field in it and no `recipient_user_profiles` field will work.  (It can ignore them in that first commit).  
* Add a new client_capability, called `stream_typing_notifications`, that is plumbed through exactly the same way that `bulk_message_deletion` is.  You can even do this as a new first commit (with it not having a functional effect in that commit, and then the main commit for typing indicators would just do that).
* Then do this backend commit, with this line having `type="typing"`, and logic in `zerver/tornado/event_queue.py` to use a `process_typing_event` little function that only adds the `typing` events that have a `stream_id` field in them if `client.stream_typing_notifications`.  It should let you simplify the API documentation changes as well.

I think with that set of changes, I'd be happy to merge this series through having the complete backend.

chdinesh1089 replied:
> logic in `zerver/tornado/event_queue.py` to use a `process_typing_event` little function that only adds the `typing` events that have a `stream_id` field in them if `client.stream_typing_notifications`.
 
 
 We could also tweak `accepts_event` return `False` in case of `stream_id` with `client.stream_typing_notifications` set to `False`. That seems better to me as `process_typing_event` does a bit of duplication from `process_event`. But for now I added `process_typing_event` as you suggested. I'd like to know what you think about the other approach and if there are any issues possible with that approach.

timabbott replied:
Yeah, I think your `accepts_event` strategy is a better way to do this.

chdinesh1089 replied:
Ok, done.

* I didn't make any commit for `server_events.js` to not crash when seeing `typing` event with a `stream_id` field but instead set the `client.stream_typing_notifications` to `False` and made it `True` in the frontend commit as the goal was to be able to merge the backend changes without any side effects and making changes to `server_events.js` felt unnecessary since we have `client.stream_typing_notifications`. 
* Added a new client_capability. This is the new first commit.
* Changed `stream_typing` event to `typing`.

timabbott commented on static/js/typing.js:
```diff
@@ -7,27 +7,53 @@ import * as channel from "./channel";
+    send_typing_notification_ajax(data);
+}
+
+function send_typing_notification_based_on_msg_type(to, operation) {
```
s/msg/message/ -- we try to avoid using abberviations.

timabbott commented on static/js/typing.js:
```diff
@@ -50,15 +76,27 @@ function get_current_time() {
+        return {stream_id, topic};
+    }
+    return null;
+}
```
The `return null` might be better written as a `throw new Error("Unexpected message type");`

chdinesh1089 replied:
Okay, changed. looks better now.

timabbott replied:
Cool; you might want the same in `send_typing_notification_based_on_message_type`?

chdinesh1089 replied:
Just realised that we do not need to throw any error in both cases:
* The `return null` in case of `get_recipient` is intended. Getting the message type to be something else than private or stream means that we aren't talking to anybody and aren't interested in sending any updates to the server. That'll be taken care of by `typing_status.update` if we send it `null` through `get_recipient`.
* I implemented `send_typing_notification_based_on_message_type` badly by using `compose_state.get_message_type()` to figure out `message_type`. This would lead to bugs. One such is when code in `typing_status` tells this function to send a stoped typing request to server, this will do nothing if the compose box was closed by user. It should instead use the existence of `stream_id` in `to`. 

Made the required changes in each case.

timabbott commented on static/js/typing_data.js:
```diff
@@ -51,11 +51,17 @@ export function get_group_typists(group) {
+export function get_all_pms_typists() {
+    let pms_typists = [];
+    for (const [key, typists] of typist_dct.entries()) {
+        if (!key.startsWith("{")) {
```
Can you add a comment on why this check will produce correct results?  It strikes me as a bit hacky.

chdinesh1089 replied:
It could've been written better as:
```
        try {
            JSON.parse(key);  // Key is a JSON string for streams typing
        }
        catch {
            pms_typists = pms_typists.concat(typists);
        }
```

Changed.

chdinesh1089 commented:
Thanks for the review @timabbott ! I’ll do the suggested changes after my exams. (It’ll be a bit late since [exams](https://www.nitandhra.ac.in/main/Announcements/2021/II%20Semester%20End%20Examinations%20Time%20Table%20for%20II,%20III%20&%20IV%20B.Tech%20students%20of%20A.Y.%202020-21.pdf) were rescheduled to end on 22nd instead of 17th.)

chdinesh1089 commented on zerver/tests/test_events.py:
```diff
@@ -708,6 +709,60 @@ def test_typing_events(self) -> None:
+        )
+        check_typing_stop("events[0]", events[0])
+
+        # Having client_capability `stream_typing_notification=False`
+        # shouldn't produce any events.
+        events = self.verify_action(
+            lambda: do_send_stream_typing_notification(
+                self.user_profile,
+                "start",
+                stream_id,
+                topic,
+            ),
+            state_change_expected=False,
+            stream_typing_notifications=False,
+            num_events=0,
+        )
+        self.assertEqual(events, [])
+
+        events = self.verify_action(
+            lambda: do_send_stream_typing_notification(
+                self.user_profile,
+                "stop",
+                stream_id,
+                topic,
+            ),
+            state_change_expected=False,
+            stream_typing_notifications=False,
+            num_events=0,
+        )
+        self.assertEqual(events, [])
```
Added these to test our new `client_capability` functionality. Is it a correct way to test something like this?

timabbott commented:
I made a few tweaks to the API documentation (critically, bumping API_FEATURE_LEVEL) and pushed back here.  I think we're close to being able to merge the backend commits.

timabbott commented:
OK, I'm satisfied with the backend commits after the following tweaks; merged them as the series ending with 
27e4f5da92305798b37aa866dfdb480005400d29 in order to get the API_FEATURE_LEVEL bump in.  Changes include:
* Added a `message_type` field in the event format, mostly for clarity.
* Added documentation of which fields are new in the `zulip.yaml` documentation; and more generally tweaked the API documentation.
* Added a check that we have permission to send messages to the stream in order to send typing notifications there.

I haven't really reviewed the frontend logic for this yet.

chdinesh1089 commented:
Nice. I'll make a few changes to the frontend to make use of the `message_type` field.

timabbott commented:
Hmm, yeah, that will surely improve things.  Also if you rebase, I already fixed the CI failure in master.

chdinesh1089 commented:
@timabbott I accidentally [removed](https://github.com/zulip/zulip/commit/27e4f5da92305798b37aa866dfdb480005400d29#diff-7d09c06f7ab16ed46eeff8eb6e1a342f75fc9f3c780c6f10643303145316581eL9808) `required: true` for `to` param to be an optional field in API. This made the [API docs show that `to` param is optional.](https://chat.zulip.org/api/set-typing-status#parameter-to) Pushed a short commit for this.

chdinesh1089 commented:
A write up as requested in https://github.com/zulip/zulip/pull/19200#issuecomment-938252625
> For the typing notifications one, I'd really appreciate your doing a bit of a writeup on the strategy for how you managed the issues around potential for duplicated code, since that's the thing I've been stuck on in thinking about that PR.

### Sending start/stop notifications

This primarily relies on `static/shared/js/typing_status.js` for the timing part i.e when (or how often) to send start/stop. `typing_status.update(worker, new_recipient)`, the only function used outside of its module, is called (from `static/js/typing.js`) every time we might want to send a notification and it takes care of the rest. The `worker` argument is expected to look as follows:
```
    const worker = {
        get_current_time,
        notify_server_start,
        notify_server_stop,
    };
```
`notify_server_start(to)` (and `_stop`) are just functions that send relevant requests taking in the recipient as `to`.

The only part that changes when we integrate stream typing notifications is, the `new_recipient`(and `to`) argument. Previously, it used an array of user IDs. Now, it can also be an object(`{stream_id, topic}`). There's no problem with `typing_status.update` to support the new possibility for `new_recipient`. So, we do not need any functional changes in `typing_status`.

But the insides of `notify_server_start/stop` would require a few `if/else`s, So I've extracted those into smaller functions. I think that'll be clear on looking at `typing.js`.

(All this is done in the first commit)

### Displaying typing... notifications for the receiver
The flow of this so far looks as follows:
1. `server_events_dispatch.js` calls typing_events.display_notification(event) upon receiving a typing notification
2. `display_notification` stores the typists into `typing_data.typist_dct` using `typing_data.add_typists(key, sender_id)`. This data is later used to actually display typists based on narrow where the `key` in `typing_data.add_typists` indicates the narrow. (for e.g. it is a string of sorted user ids for group pms)
3. Actually display typing notifications using `typing_events.render_notifications_for_narrow()`. This also takes care of hiding notifications. This function also has some calls in narrow.js to make sure these notifications are displayed correctly when a narrow is changed.
4. Start timer to hide the notification `hide_notification` which also takes care of removing typists from `typing_data.typist_dct` using `typing_data.remove_typists(key, sender_id)`.

The flow is the same for stream typing notifications too but now we want to store `{stream_id, topic}` along with typists in `typing_data.typist_dct` but that wouldn't be cleaner with two possible forms of `key` value when we rely on the key format at a few other places for fetching typists. So I added `pm_typists_dict`, and `stream_typists_dict` and removed the earlier one. The `add_typist(key, typist)` now becomes `add_typist(key, typist, message_type)` to separately store the typists into their respective dicts. Similarly for `remove_typist` too.  The `message_type` is sent using `event.message_type` in `typing_events.js`.

`typist_data.js` also has functions for proper generation of keys for each type of message: `get_pms_key`, `get_topic_key`. These are used in a bunch of places to get items from the two dicts. I recommend checking their references.

`typist_data.js` also has `inbound_timer_dict` which keeps track of timer ids of `setTimeout`s to remove typists. Using this for both stream and private notifications wouldn't do any harm and keep things simple. Moreover, the wrapper functions clearing these receive the key. So, writing if/else here by creating two objects each for stream and private will only complicate things.

pdurbin commented:
@chdinesh1089 hi! I'd love to see typing indicators for streams and it seems like this PR of yours gets us there! Thanks! ❤️ 

I noticed a related topic in chat and [asked](https://chat.zulip.org/#narrow/stream/101-design/topic/typing.20indicators.20and.20read.20receipts/near/1526697) about this PR there. It [sounds like](https://chat.zulip.org/#narrow/stream/101-design/topic/typing.20indicators.20and.20read.20receipts/near/1527159) the next step is to rebase. I just thought I'd pass that along. Have a great weekend! 🏖️

brijsiyag commented:
In the event that @chdinesh1089 has no intention of working on it, I would be interested in taking over. Thank you!!

pdurbin commented:
@brijsiyag great! Locally I rebased `main` as of b21d93c into this branch and the merge conflicts don't look too bad. There were two auto merges and one test file that needs to be fixed manually:

```
Auto-merging web/src/typing.js
Auto-merging web/tests/typing_status.test.js
CONFLICT (content): Merge conflict in web/tests/typing_status.test.js
```

Here are the two places where `typing_status.test.js` has conflicts, in only a few lines:

<img width="1114" alt="Screenshot 2023-03-19 at 9 00 10 PM" src="https://user-images.githubusercontent.com/21006/226223229-e769eb00-31e4-47d5-b2be-43ecede36282.png">

<img width="1114" alt="Screenshot 2023-03-19 at 9 00 31 PM" src="https://user-images.githubusercontent.com/21006/226223240-d83a9f7e-b524-40ff-8e97-060b77ed7034.png">

I hope this helps either you or @chdinesh1089. ❤️

I haven't set up a dev environment yet, but maybe I can help with testing or something. 😅

chdinesh1089 commented:
Thanks for bumping this thread, @pdurbin!

I haven't been getting much free time these days. I might delay this if I pick up again. I'd be super happy to see this feature merged soon so I'd appreciate if @brijsiyag or anyone would like to take this forward. 🙂

@pdurbin it's quite simple to setup development environment. see https://zulip.readthedocs.io/en/latest/development/overview.html

pdurbin commented:
@chdinesh1089 sure. Coming from Slack and Matrix, I'm interested in this feature.

@brijsiyag are you interested in taking over? I do have a dev environment set up now but I'm having trouble resolving all the merge conflicts.

brijsiyag commented:
> @brijsiyag are you interested in taking over?

Yes @pdurbin , I'm working on this.

timabbott commented:
Closing in favor of https://github.com/zulip/zulip/pull/26904, thanks for all the work on this @chdinesh1089!
